### ISSUE TITLE
Feature request: ability to change the request asynchronously in the RequestAdapter

### ISSUE THREAD

**Issue by ricardopereira** [Alamofire/Alamofire#2517]
* [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/Alamofire/Alamofire/blob/master/CONTRIBUTING.md).

## What did you do?

As I mentioned [here](https://github.com/Alamofire/Alamofire/pull/1450#issuecomment-389277756), I want to request a new access token in my `RequestAdapter` implementation when the token has expired. I know in advance when will the access token expire so, I think it would be nice to spare a request that will probably answer with a 401.

## What did you expect to happen?

I was expecting something like:

```
typealias RequestAdaptCompletion = (_ urlRequest: URLRequest) -> Void

func adapt(_ urlRequest: URLRequest, completion: @escaping RequestAdaptCompletion) throws {
    if userTokenDetails.expires.timeIntervalSince(userTokenDetails.issued) <= 0 {
        reauthorize { newUserTokenDetails in
            urlRequest.setValue("Bearer " + newUserTokenDetails.accessToken, forHTTPHeaderField: "Authorization")
            completion(urlRequest)
        }
    }
    else {
        completion(urlRequest)
    }
}
```

> @ricardopereira It would be much easier to just allow adapter errors to trigger retry behavior. I'm not sure if that would work now, but it might become possible for Alamofire 5. Such a feature likely wouldn't ship until then anyway.

but as @jshier mentioned in #1450, maybe triggering an error in the Adapter and call retry should be enough.

## Alamofire Environment

**Alamofire version:** 4.7.2
**Xcode version:** 9.3.1
**Swift version:** 4.0
**Platform(s) running Alamofire:** iOS
**macOS version running Xcode:** High Sierra 10.13.4 (17E202)

## Demo Project

No demo but I can arrange something if needed.

🤩 Thank you for your amazing work!

**Comment by jshier** [comment]
While I'm pretty sure this will work as @cnoon described previously, you can also try the `alamofire5-core-rewrite` branch and see what behavior you get there. `AdaptError` has been removed, you should check for your adapter errors in the retrier before deciding to retry.

**Comment by FrancoSabadini** [comment]
Another case to make for this is that sometimes we are using 3rd party libraries for authorization which only provide async methods to get the access token (e.g., Firebase) and we can't use the `RequestAdapter` to add the access token to all the requests, nor can we change the third party library APIs. This feature would be great for this.

**Comment by TwunTee** [comment]
@FrancoSabadini how are you sending the firebase token. I am using getIdToken() and if that fails i am using the retrier, do you have any better solution.

**Comment by FrancoSabadini** [comment]
Hi @TwunTee, I'm using `getIDTokenResult()`, but it receives a closure, it doesn't return the result synchronously. This is from the Firebase SDK code: 

```
- (void)getIDTokenResultForcingRefresh:(BOOL)forceRefresh
                            completion:(nullable FIRAuthTokenResultCallback)completion
```

From what I see, `getIdToken()` also returns the token in a closure:

```
- (void)getIDTokenWithCompletion:(nullable FIRAuthTokenCallback)completion
    NS_SWIFT_NAME(getIDToken(completion:));
```

Is the method you are using synchronous? What version of the Firebase SDK are you using?

**Comment by TwunTee** [comment]
@FrancoSabadini  I am using the same method too. I am doing it as part of the retrier which accepts a completion block

**Comment by jshier** [comment]
This would actually be pretty trivial to implement for Alamofire 5, given the entire request creation process is now async. It just depends on whether the API change to `RequestAdapter` is worth it.

**Comment by cnoon** [comment]
I agree here with @FrancoSabadini that we should look to make the adapter async in AF5. It was always a bit clunky in AF4 to have to force the request to have to fail before you could fetch a new token. It was the only viable approach at the time though given that request creation was synchronous. Now that it is async, the adaptation process could be smoothed out by making it async to match the retry API.

This will make certain cases a bit more complex though, but I think it's worth the tradeoff. Now you'll have to handle the case where you could be managing a queue of adapting requests as well as retrying requests. You'll also have to handle the case where both adapt and retry could be trying to refresh the credential at the same time. However, I don't think this adds a huge amount of complexity, just something to be away of.

**Comment by jshier** [comment]
Looking into implementing async `RequestAdapter`s now.

**Comment by jshier** [comment]
Making `RequestAdapter` async would require changing the function from:
```swift
func adapt(_ urlRequest: URLRequest) throws -> URLRequest
```
to:
```swift
func adapt(_ urlRequest: URLRequest, completion: @escaping (_ result: Result<URLRequest) -> Void)
```
due to the need to preserve the error. This could be rather awkward, but I'll see how it feels.

**Comment by jshier** [comment]
@ricardopereira @FrancoSabadini Take a look at #2628 and see what you think.

**Comment by ricardopereira** [comment]
@jshier It looks good to me. Thanks!

**Comment by FrancoSabadini** [comment]
Looks good to me too @jshier, thanks!! :)
